packages/vuetify/src/composables/theme.ts

```diff
@@ -0,0 +1,260 @@
+  defaultTheme: Ref<string>
+  setTheme: (key: string, theme: ThemeOption) => void
+}
+
+interface ThemeProvide {
+  themeClass: Ref<string>
+  current: Ref<string>
+  next: () => void
+  prev: () => void
+  setTheme: (key: string, theme: ThemeOption) => void
+}
+
+const VuetifyThemeSymbol: InjectionKey<ThemeProvide> = Symbol.for('vuetify:theme')
+
+const step = <T>(arr: T[], current: T, steps: number): T => arr[(arr.indexOf(current) + arr.length + steps) % arr.length]
+
+export const provideTheme = (props: { theme?: string } = {}): ThemeProvide => {
+  const vuetify = useVuetify()
+  const themeProvide = inject(VuetifyThemeSymbol, null)
+
+  const internal = ref<string | null>(null)
+  const current = computed<string>({
+    get: () => {
+      return props.theme ?? internal.value ?? themeProvide?.current.value ?? vuetify.theme.defaultTheme.value
+    },
+    set (theme: string) {
+      internal.value = theme
+    },
+  })
+
+  const themeClass = computed(() => `theme--${current.value}`)
+
+  const next = () => current.value = step(Object.keys(vuetify.theme.themes.value), current.value, 1)
+  const prev = () => current.value = step(Object.keys(vuetify.theme.themes.value), current.value, -1)
+
+  const newThemeProvide: ThemeProvide = {
+    themeClass,
+    current,
+    next,
+    prev,
+    setTheme: vuetify.theme.setTheme,
+  }
+
+  provide(VuetifyThemeSymbol, newThemeProvide)
+
+  return newThemeProvide
+}
+
+export const useTheme = () => {
+  const theme = inject(VuetifyThemeSymbol)
+
+  if (!theme) throw new Error('Could not find vuetify theme provider')
+
+  return theme
+}
+
+const defaultThemeOptions: ThemeOptions = {
+  defaultTheme: 'light',
+  variations: {
+    darken: 1,
+    lighten: 2,
+  },
+  themes: {
+    light: {
+      background: '#eeeeee',
+      surface: '#aaaaaa',
+      primary: '#1976D2',
+      secondary: '#424242',
+      accent: '#82B1FF',
+      error: '#FF5252',
+      info: '#2196F3',
+      success: '#4CAF50',
+      warning: '#FB8C00',
+    },
+    dark: {
+      background: '#555555',
+      surface: '#333333',
+      primary: '#2196F3',
+      secondary: '#424242',
+      accent: '#FF4081',
+      error: '#FF5252',
+      info: '#2196F3',
+      success: '#4CAF50',
+      warning: '#FB8C00',
+    },
+  },
+}
+
+export const createTheme = (options?: ThemeOptions): ThemeInstance => {
+  const combinedOptions = {
+    ...defaultThemeOptions,
+    ...(options as InternalThemeOptions),
+  }
+  const styleEl = ref<HTMLStyleElement | undefined>()
+  const defaultTheme = ref<string>(combinedOptions.defaultTheme)
+  const themes = ref<Record<string, ThemeOption>>(combinedOptions.themes)
+  const variations = ref(combinedOptions.variations)
+
+  const toHex = (v: number) => `#${v.toString(16).repeat(3)}`
+
+  const genOnColor = (color: string) => {
+    // naive solution
+    const goal = 8
+    let curr = 0
+    while (contrastRatio(color, toHex(curr)) < goal && curr < 255) {
```

Comment:
Last time is tried to implement this: 414cfe1d52f9cb8d45009cbc844efb71e72a299b

I don't recall where the 0.18 value came from.